src/azul/terra.py

```diff
@@ -484,14 +484,21 @@ def snapshot_names_by_id(self) -> Dict[str, str]:
         """
         limit = 100
         endpoint = self._repository_endpoint('snapshots')
-        response = self._request('GET', endpoint, fields=dict(limit=str(limit)))
-        response = self._check_response(endpoint, response)
-        # FIXME: Implement paging
-        #        https://github.com/DataBiosphere/azul/issues/3093
-        require(response['total'] <= limit, 'Too many snapshots', response['total'])
+        snapshots = []
+        while True:
+            response = self._request('GET', endpoint, fields={
+                'offset': len(snapshots),
+                'limit': limit
+            })
+            response = self._check_response(endpoint, response)
+            total = response['total']
+            snapshots += response['items']
+            if len(snapshots) >= total:
+                assert len(snapshots) == total
+                break
```

Comment:
I don't think it's appropriate to use `assert` here. Hannes is strict about not using `assert` if the assertion could be triggered by an external input. If there was a bug in TDR that caused them to send us a response where there was a mismatch between the reported total and the actual number of snapshots, that would trigger this assertion despite there not being any logical errors in our own code. As such I think this would be better addressed using `reject`/`require` instead of `assert`.